cmd/azure.go

```diff
@@ -0,0 +1,363 @@
+	return StorageInfo{}
+}
+
+// MakeBucket - Create bucket.
+func (a AzureObjects) MakeBucket(bucket string) error {
+	return azureToObjectError(traceError(a.client.CreateContainer(bucket, storage.ContainerAccessTypePrivate)), bucket)
+}
+
+// GetBucketInfo - Get bucket metadata.
+func (a AzureObjects) GetBucketInfo(bucket string) (BucketInfo, error) {
+	// Azure does not have an equivalent call, hence use ListContainers.
+	resp, err := a.client.ListContainers(storage.ListContainersParameters{
+		Prefix: bucket,
+	})
+	if err != nil {
+		return BucketInfo{}, azureToObjectError(traceError(err), bucket)
+	}
+	for _, container := range resp.Containers {
+		if container.Name == bucket {
+			t, e := time.Parse(time.RFC1123, container.Properties.LastModified)
+			if e != nil {
+				continue
+			}
+			return BucketInfo{
+				Name:    bucket,
+				Created: t,
+			}, nil
+		}
+	}
+	return BucketInfo{}, traceError(BucketNotFound{})
+}
+
+// ListBuckets - Use Azure equivalent ListContainers.
+func (a AzureObjects) ListBuckets() (buckets []BucketInfo, err error) {
+	resp, err := a.client.ListContainers(storage.ListContainersParameters{})
+	if err != nil {
+		return nil, azureToObjectError(traceError(err))
+	}
+	for _, container := range resp.Containers {
+		t, e := time.Parse(time.RFC1123, container.Properties.LastModified)
+		if e != nil {
+			return nil, traceError(e)
+		}
+		buckets = append(buckets, BucketInfo{
+			Name:    container.Name,
+			Created: t,
+		})
+	}
+	return buckets, nil
+}
+
+// DeleteBucket - Use Azure equivalent DeleteContainer.
+func (a AzureObjects) DeleteBucket(bucket string) error {
+	return azureToObjectError(traceError(a.client.DeleteContainer(bucket)), bucket)
+}
+
+// ListObjects - Use Azure equivalent ListBlobs.
+func (a AzureObjects) ListObjects(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	resp, err := a.client.ListBlobs(bucket, storage.ListBlobsParameters{
+		Prefix:     prefix,
+		Marker:     marker,
+		Delimiter:  delimiter,
+		MaxResults: uint(maxKeys),
+	})
+	if err != nil {
+		return result, azureToObjectError(traceError(err), bucket, prefix)
+	}
+	result.IsTruncated = resp.NextMarker != ""
+	result.NextMarker = resp.NextMarker
+	for _, object := range resp.Blobs {
+		t, e := time.Parse(time.RFC1123, object.Properties.LastModified)
+		if e != nil {
+			continue
+		}
+		result.Objects = append(result.Objects, ObjectInfo{
+			Bucket:          bucket,
+			Name:            object.Name,
+			ModTime:         t,
+			Size:            object.Properties.ContentLength,
+			MD5Sum:          object.Properties.Etag,
+			ContentType:     object.Properties.ContentType,
+			ContentEncoding: object.Properties.ContentEncoding,
+		})
+	}
+	result.Prefixes = resp.BlobPrefixes
+	return result, nil
+}
+
+// GetObject - Use Azure equivalent GetBlobRange.
+func (a AzureObjects) GetObject(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error) {
```

Comment:
this is fine, we inherit the limitations of azure and its SDK. we'll be a minimal layer on top of azure unless it becomes a concern later that needs to be fixed